Remotion/ObjectBinding/Web/UI/Controls/Validation/BocListValidationFailureRepository.cs

```diff
@@ -0,0 +1,315 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// The re-motion Core Framework is free software; you can redistribute it
+// and/or modify it under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation; either version 2.1 of the
+// License, or (at your option) any later version.
+//
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Remotion.ObjectBinding.Validation;
+using Remotion.Utilities;
+
+namespace Remotion.ObjectBinding.Web.UI.Controls.Validation
+{
+  /// <summary>
+  /// Repository which contains all <see cref="BusinessObjectValidationFailure"/> belonging to a BocList.
+  /// </summary>
+  public class BocListValidationFailureRepository: IBocListValidationFailureRepository
+  {
+    private class HandleableFailure
+    {
+      public BusinessObjectValidationFailure Failure { get; }
+
+      public bool Handled { get; private set; }
+
+      public HandleableFailure (BusinessObjectValidationFailure failure)
+      {
+        Failure = failure;
+        Handled = false;
+      }
+
+      public void MarkAsHandled ()
+      {
+        Handled = true;
+      }
+    }
+
+    private class RowFailureContainer
+    {
+      private List<HandleableFailure>? _rowFailures;
+      private Dictionary<IBocColumnDefinitionWithValidationSupport, List<HandleableFailure>>? _cellFailures;
+
+      public void AddRowFailures (IEnumerable<BusinessObjectValidationFailure> failures)
+      {
+        _rowFailures ??= new List<HandleableFailure>();
+        _rowFailures.AddRange(failures.Select(e => new HandleableFailure(e)));
+      }
+
+      public void AddCellFailures (IBocColumnDefinitionWithValidationSupport cellDefinition, IEnumerable<BusinessObjectValidationFailure> failures)
+      {
+        List<HandleableFailure> cellFailures;
+        _cellFailures ??= new Dictionary<IBocColumnDefinitionWithValidationSupport, List<HandleableFailure>>();
+        if (_cellFailures.TryGetValue(cellDefinition, out var existingFailureList))
+        {
+          cellFailures = existingFailureList;
+        }
+        else
+        {
+          cellFailures = new List<HandleableFailure>();
+          _cellFailures[cellDefinition] = cellFailures;
+        }
+
+        cellFailures.AddRange(failures.Select(e => new HandleableFailure(e)));
+      }
+
+      public IReadOnlyCollection<HandleableFailure> GetRowAndCellFailures ()
```

Comment:
please check, I think we can replace all the IReadOnlyCollection<> types with IEnumerable<> 

This allows us to use Concat in this method, preventing the copy operation. And since we're iterating the results anyway, we can get rid of all the listbuilding and make things completely lazy until returned from the outer/public api.